File: phases/16-software-engineering-and-architecture/05-modern-patterns-functional-core-imperative-shell/code/main.ts
// Modern Patterns — Functional Core / Imperative Shell
// Phase 16 — Software Engineering & Architecture
//
// Demonstrates separating pure business logic from side effects
// in an order processing system using TypeScript.

// ============================================================================
// FUNCTIONAL CORE — Pure functions, no side effects, no dependencies
// ============================================================================

interface OrderItem {
  productId: string;
  name: string;
  quantity: number;
  unitPrice: number;
}

interface Order {
  customerId: string;
  items: OrderItem[];
}

interface ValidatedOrder {
  order: Order;
}

interface PricedOrder {
  order: Order;
  subtotal: number;
  discountAmount: number;
  total: number;
  discountReason: string;
}

enum CustomerTier {
  Gold = "gold",
  Silver = "silver",
  Bronze = "bronze",
}

interface Discount {
  amount: number;
  reason: string;
}

type ValidationError =
  | { kind: "empty_order" }
  | { kind: "missing_customer_id" }
  | { kind: "negative_quantity"; productId: string };

type OrderDecision =
  | { kind: "accept"; pricedOrder: PricedOrder }
  | { kind: "reject"; error: ValidationError };

function validateOrder(order: Order): ValidatedOrder {
  if (!order.customerId || order.customerId.trim().length === 0) {
    throw { kind: "missing_customer_id" } as ValidationError;
  }
  if (order.items.length === 0) {
    throw { kind: "empty_order" } as ValidationError;
  }
  for (const item of order.items) {
    if (item.quantity <= 0) {
      throw { kind: "negative_quantity", productId: item.productId } as ValidationError;
    }
  }
  return { order };
}

function calculateDiscount(order: ValidatedOrder, tier: CustomerTier): Discount {
  const subtotal = order.order.items.reduce(
    (sum, item) => sum + item.quantity * item.unitPrice,
    0
  );
  const rates: Record<CustomerTier, { pct: number; label: string }> = {
    [CustomerTier.Gold]: { pct: 0.15, label: "gold_tier" },
    [CustomerTier.Silver]: { pct: 0.10, label: "silver_tier" },
    [CustomerTier.Bronze]: { pct: 0.05, label: "bronze_tier" },
  };
  const { pct, label } = rates[tier];
  return { amount: subtotal * pct, reason: label };
}

function applyDiscount(order: ValidatedOrder, discount: Discount): PricedOrder {
  const subtotal = order.order.items.reduce(
    (sum, item) => sum + item.quantity * item.unitPrice,
    0
  );
  const total = Math.max(0, subtotal - discount.amount);
  return {
    order: order.order,
    subtotal,
    discountAmount: discount.amount,
    total,
    discountReason: discount.reason,
  };
}

function processOrderCore(order: Order, tier: CustomerTier): OrderDecision {
  try {
    const validated = validateOrder(order);
    const discount = calculateDiscount(validated, tier);
    const pricedOrder = applyDiscount(validated, discount);
    return { kind: "accept", pricedOrder };
  } catch (e) {
    return { kind: "reject", error: e as ValidationError };
  }
}

// ============================================================================
// IMPERATIVE SHELL — I/O, database, HTTP, logging
// ============================================================================

interface ShellConfig {
  dbUrl: string;
  logLevel: string;
}

async function shellFetchCustomerTier(
  customerId: string,
  config: ShellConfig
): Promise<CustomerTier> {
  console.log(
    `[SHELL][db] Querying tier for customer '${customerId}' at ${config.dbUrl}`
  );
  // Simulate async database call
  await new Promise((resolve) => setTimeout(resolve, 10));
  if (customerId.startsWith("G")) return CustomerTier.Gold;
  if (customerId.startsWith("S")) return CustomerTier.Silver;
  return CustomerTier.Bronze;
}

async function shellSaveOrder(pricedOrder: PricedOrder, config: ShellConfig): Promise<void> {
  console.log(
    `[SHELL][db] Saving order for customer '${pricedOrder.order.customerId}' — total: $${pricedOrder.total.toFixed(2)} at ${config.dbUrl}`
  );
  await new Promise((resolve) => setTimeout(resolve, 10));
}

async function shellSendNotification(pricedOrder: PricedOrder): Promise<void> {
  console.log(
    `[SHELL][http] POST /notify — customer ${pricedOrder.order.customerId} charged $${pricedOrder.total.toFixed(2)} ` +
      `(discount: $${pricedOrder.discountAmount.toFixed(2)} — ${pricedOrder.discountReason})`
  );
  await new Promise((resolve) => setTimeout(resolve, 10));
}

function shellLogRejection(error: ValidationError): void {
  console.log(`[SHELL][log] Order rejected: ${JSON.stringify(error)}`);
}

async function shellRun(order: Order, config: ShellConfig): Promise<void> {
  console.log(
    `[SHELL] --- Processing order for customer '${order.customerId}' ---`
  );

  const tier = await shellFetchCustomerTier(order.customerId, config);

  const decision = processOrderCore(order, tier);

  switch (decision.kind) {
    case "accept": {
      await shellSaveOrder(decision.pricedOrder, config);
      await shellSendNotification(decision.pricedOrder);
      console.log(
        `[SHELL] ✓ Order accepted — subtotal: $${decision.pricedOrder.subtotal.toFixed(2)}, ` +
          `discount: $${decision.pricedOrder.discountAmount.toFixed(2)}, ` +
          `total: $${decision.pricedOrder.total.toFixed(2)}`
      );
      break;
    }
    case "reject": {
      shellLogRejection(decision.error);
      console.log("[SHELL] ✗ Order rejected");
      break;
    }
  }
}

// ============================================================================
// TESTS — Pure core tests require NO infrastructure, NO mocks
// ============================================================================

function assert(condition: boolean, message: string): void {
  if (!condition) throw new Error(`Assertion failed: ${message}`);
}

function assertEqual<T>(actual: T, expected: T, label: string): void {
  assert(JSON.stringify(actual) === JSON.stringify(expected), `${label}: ${JSON.stringify(actual)} !== ${JSON.stringify(expected)}`);
}

function runCoreTests(): void {
  const sampleOrder: Order = {
    customerId: "CUST-001",
    items: [
      { productId: "SKU-A", name: "Widget", quantity: 3, unitPrice: 10.0 },
      { productId: "SKU-B", name: "Gadget", quantity: 1, unitPrice: 25.0 },
    ],
  };

  // validate — valid order
  const validated = validateOrder(sampleOrder);
  assertEqual(validated.order, sampleOrder, "validate valid order");

  // validate — empty customer ID
  let caught = false;
  try {
    validateOrder({ ...sampleOrder, customerId: "" });
  } catch (e) {
    caught = true;
    assert((e as ValidationError).kind === "missing_customer_id", "empty customer id");
  }
  assert(caught, "should throw for empty customer id");

  // validate — empty items
  caught = false;
  try {
    validateOrder({ customerId: "CUST-001", items: [] });
  } catch (e) {
    caught = true;
    assert((e as ValidationError).kind === "empty_order", "empty order");
  }
  assert(caught, "should throw for empty order");

  // calculate discount — gold tier (15%)
  const goldDiscount = calculateDiscount(validated, CustomerTier.Gold);
  assert(Math.abs(goldDiscount.amount - 8.25) < 0.001, `gold discount: ${goldDiscount.amount}`);
  assertEqual(goldDiscount.reason, "gold_tier", "gold reason");

  // calculate discount — silver tier (10%)
  const silverDiscount = calculateDiscount(validated, CustomerTier.Silver);
  assert(Math.abs(silverDiscount.amount - 5.5) < 0.001, `silver discount: ${silverDiscount.amount}`);

  // apply discount — total never negative
  const megaDiscount: Discount = { amount: 99999, reason: "mega_sale" };
  const pricedWithMega = applyDiscount(validated, megaDiscount);
  assert(pricedWithMega.total === 0, "total should be clamped to 0");

  // process order — accept path
  const acceptDecision = processOrderCore(sampleOrder, CustomerTier.Gold);
  assert(acceptDecision.kind === "accept", "should accept valid order");
  if (acceptDecision.kind === "accept") {
    assert(Math.abs(acceptDecision.pricedOrder.total - 46.75) < 0.01, "total should be 46.75");
  }

  // process order — reject path
  const rejectDecision = processOrderCore(
    { customerId: "CUST-001", items: [] },
    CustomerTier.Gold
  );
  assert(rejectDecision.kind === "reject", "should reject empty order");

  console.log("[CORE TESTS] All core tests passed ✓");
}

// ============================================================================
// MAIN — Shell orchestrates the whole flow
// ============================================================================

async function main(): Promise<void> {
  // Run pure core tests first — no infrastructure needed
  runCoreTests();
  console.log();

  const config: ShellConfig = {
    dbUrl: "postgres://localhost/orders",
    logLevel: "info",
  };

  // Happy path
  const order: Order = {
    customerId: "GOLD-CUST-42",
    items: [
      { productId: "SKU-WIDGET", name: "Premium Widget", quantity: 5, unitPrice: 20.0 },
      { productId: "SKU-GADGET", name: "Super Gadget", quantity: 2, unitPrice: 50.0 },
    ],
  };

  await shellRun(order, config);

  console.log();

  // Sad path — invalid order
  const badOrder: Order = {
    customerId: "",
    items: [],
  };
  await shellRun(badOrder, config);
}

main().catch(console.error);

File: phases/16-software-engineering-and-architecture/05-modern-patterns-functional-core-imperative-shell/code/main.rs
//! Modern Patterns — Functional Core / Imperative Shell
//! Phase 16 — Software Engineering & Architecture
//!
//! Demonstrates separating pure business logic (functional core) from
//! side effects (imperative shell) in an order processing system.

use std::fmt;

// ============================================================================
// FUNCTIONAL CORE — Pure functions, no side effects, no dependencies
// ============================================================================

#[derive(Debug, Clone, PartialEq)]
struct Order {
    customer_id: String,
    items: Vec<OrderItem>,
}

#[derive(Debug, Clone, PartialEq)]
struct OrderItem {
    product_id: String,
    name: String,
    quantity: u32,
    unit_price: f64,
}

#[derive(Debug, Clone, PartialEq)]
struct ValidatedOrder {
    order: Order,
}

#[derive(Debug, Clone, PartialEq)]
struct PricedOrder {
    order: Order,
    subtotal: f64,
    discount_amount: f64,
    total: f64,
    discount_reason: String,
}

#[derive(Debug, Clone, Copy, PartialEq)]
enum CustomerTier {
    Gold,
    Silver,
    Bronze,
}

#[derive(Debug, Clone, PartialEq)]
struct Discount {
    amount: f64,
    reason: String,
}

impl Discount {
    fn new(amount: f64, reason: &str) -> Self {
        Discount {
            amount,
            reason: reason.to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
enum ValidationError {
    EmptyOrder,
    MissingCustomerId,
    NegativeQuantity { product_id: String },
}

#[derive(Debug, Clone, PartialEq)]
enum OrderDecision {
    Accept(PricedOrder),
    Reject(ValidationError),
}

impl Order {
    fn subtotal(&self) -> f64 {
        self.items.iter().map(|i| i.quantity as f64 * i.unit_price).sum()
    }
}

fn validate_order(order: &Order) -> Result<ValidatedOrder, ValidationError> {
    if order.customer_id.is_empty() {
        return Err(ValidationError::MissingCustomerId);
    }
    if order.items.is_empty() {
        return Err(ValidationError::EmptyOrder);
    }
    for item in &order.items {
        if item.quantity == 0 {
            return Err(ValidationError::NegativeQuantity {
                product_id: item.product_id.clone(),
            });
        }
    }
    Ok(ValidatedOrder {
        order: order.clone(),
    })
}

fn calculate_discount(order: &ValidatedOrder, tier: CustomerTier) -> Discount {
    let base = order.order.subtotal();
    match tier {
        CustomerTier::Gold => Discount::new(base * 0.15, "gold_tier"),
        CustomerTier::Silver => Discount::new(base * 0.10, "silver_tier"),
        CustomerTier::Bronze => Discount::new(base * 0.05, "bronze_tier"),
    }
}

fn apply_discount(order: &ValidatedOrder, discount: &Discount) -> PricedOrder {
    let subtotal = order.order.subtotal();
    let total = (subtotal - discount.amount).max(0.0);
    PricedOrder {
        order: order.order.clone(),
        subtotal,
        discount_amount: discount.amount,
        total,
        discount_reason: discount.reason.clone(),
    }
}

fn process_order_core(order: &Order, tier: CustomerTier) -> OrderDecision {
    match validate_order(order) {
        Ok(validated) => {
            let discount = calculate_discount(&validated, tier);
            let priced = apply_discount(&validated, &discount);
            OrderDecision::Accept(priced)
        }
        Err(e) => OrderDecision::Reject(e),
    }
}

// ============================================================================
// IMPERATIVE SHELL — I/O, database, HTTP, logging
// ============================================================================

struct ShellConfig {
    db_url: String,
    log_level: String,
}

fn shell_fetch_customer_tier(customer_id: &str, config: &ShellConfig) -> CustomerTier {
    println!("[SHELL][db] Querying tier for customer '{}' at {}", customer_id, config.db_url);
    if customer_id.starts_with("G") {
        CustomerTier::Gold
    } else if customer_id.starts_with("S") {
        CustomerTier::Silver
    } else {
        CustomerTier::Bronze
    }
}

fn shell_save_order(order: &PricedOrder, config: &ShellConfig) {
    println!(
        "[SHELL][db] Saving order for customer '{}' — total: ${:.2} at {}",
        order.order.customer_id, order.total, config.db_url
    );
}

fn shell_send_notification(order: &PricedOrder) {
    println!(
        "[SHELL][http] POST /notify — customer {} charged ${:.2} (discount: ${:.2} — {})",
        order.order.customer_id, order.total, order.discount_amount, order.discount_reason
    );
}

fn shell_log_rejection(error: &ValidationError) {
    println!("[SHELL][log] Order rejected: {:?}", error);
}

fn shell_run(order: &Order, config: &ShellConfig) {
    println!("[SHELL] --- Processing order for customer '{}' ---", order.customer_id);

    let tier = shell_fetch_customer_tier(&order.customer_id, config);

    let decision = process_order_core(order, tier);

    match decision {
        OrderDecision::Accept(priced) => {
            shell_save_order(&priced, config);
            shell_send_notification(&priced);
            println!(
                "[SHELL] ✓ Order accepted — subtotal: ${:.2}, discount: ${:.2}, total: ${:.2}",
                priced.subtotal, priced.discount_amount, priced.total
            );
        }
        OrderDecision::Reject(ref err) => {
            shell_log_rejection(err);
            println!("[SHELL] ✗ Order rejected");
        }
    }
}

// ============================================================================
// TESTS — Pure core tests require NO infrastructure
// ============================================================================

#[cfg(test)]
mod core_tests {
    use super::*;

    fn sample_order() -> Order {
        Order {
            customer_id: "CUST-001".to_string(),
            items: vec![
                OrderItem {
                    product_id: "SKU-A".to_string(),
                    name: "Widget".to_string(),
                    quantity: 3,
                    unit_price: 10.0,
                },
                OrderItem {
                    product_id: "SKU-B".to_string(),
                    name: "Gadget".to_string(),
                    quantity: 1,
                    unit_price: 25.0,
                },
            ],
        }
    }

    #[test]
    fn test_validate_valid_order() {
        let order = sample_order();
        let result = validate_order(&order);
        assert!(result.is_ok());
        assert_eq!(result.unwrap().order, order);
    }

    #[test]
    fn test_validate_empty_customer_id() {
        let mut order = sample_order();
        order.customer_id = String::new();
        assert_eq!(validate_order(&order), Err(ValidationError::MissingCustomerId));
    }

    #[test]
    fn test_validate_empty_order() {
        let order = Order {
            customer_id: "CUST-001".to_string(),
            items: vec![],
        };
        assert_eq!(validate_order(&order), Err(ValidationError::EmptyOrder));
    }

    #[test]
    fn test_calculate_discount_gold() {
        let order = sample_order();
        let validated = validate_order(&order).unwrap();
        let discount = calculate_discount(&validated, CustomerTier::Gold);
        assert!((discount.amount - 8.25).abs() < 0.001);
        assert_eq!(discount.reason, "gold_tier");
    }

    #[test]
    fn test_calculate_discount_silver() {
        let order = sample_order();
        let validated = validate_order(&order).unwrap();
        let discount = calculate_discount(&validated, CustomerTier::Silver);
        assert!((discount.amount - 5.50).abs() < 0.001);
    }

    #[test]
    fn test_apply_discount_total_never_negative() {
        let order = sample_order();
        let validated = validate_order(&order).unwrap();
        let discount = Discount::new(99999.0, "mega_sale");
        let priced = apply_discount(&validated, &discount);
        assert!((priced.total - 0.0).abs() < 0.001);
    }

    #[test]
    fn test_process_order_accept() {
        let order = sample_order();
        let decision = process_order_core(&order, CustomerTier::Gold);
        match decision {
            OrderDecision::Accept(priced) => {
                assert!((priced.total - 46.75).abs() < 0.01);
            }
            OrderDecision::Reject(_) => panic!("Expected Accept"),
        }
    }

    #[test]
    fn test_process_order_reject_empty() {
        let order = Order {
            customer_id: "CUST-001".to_string(),
            items: vec![],
        };
        let decision = process_order_core(&order, CustomerTier::Gold);
        assert!(matches!(decision, OrderDecision::Reject(ValidationError::EmptyOrder)));
    }
}

// ============================================================================
// MAIN — Shell orchestrates the whole flow
// ============================================================================

fn main() {
    let config = ShellConfig {
        db_url: "postgres://localhost/orders".to_string(),
        log_level: "info".to_string(),
    };

    let order = Order {
        customer_id: "GOLD-CUST-42".to_string(),
        items: vec![
            OrderItem {
                product_id: "SKU-WIDGET".to_string(),
                name: "Premium Widget".to_string(),
                quantity: 5,
                unit_price: 20.0,
            },
            OrderItem {
                product_id: "SKU-GADGET".to_string(),
                name: "Super Gadget".to_string(),
                quantity: 2,
                unit_price: 50.0,
            },
        ],
    };

    shell_run(&order, &config);

    println!();

    let bad_order = Order {
        customer_id: String::new(),
        items: vec![],
    };
    shell_run(&bad_order, &config);
}